src/pipeline/mod.rs

```diff
@@ -0,0 +1,213 @@
+#![cfg(feature = "unstable")]
+
+use futures::lazy;
+use prelude::*;
+use iter::internal::*;
+use std::collections::VecDeque;
+use Scope;
+
+mod test;
+
+pub trait IntoParPipeline
+    where Self: Sized + IntoIterator, Self::Item: Send,
+{
+    /// Converts a sequential iterator into a parallel one. The
+    /// sequential iterator will be driven from a single thread, and a
+    /// Rayon task will be created for each item in the iterator.
+    ///
+    /// In general, it is preferable to use a native parallel iterator
+    /// if they are available. It may also be better to collect the
+    /// items into a vector and then use `into_par_iter()` on that
+    /// vector -- this can help us to create fewer parallel tasks and
+    /// reduce overall overhead.
+    ///
+    /// If the sequential iterator is not `Send`, see `into_par_pipeline_scoped()`.
+    fn into_par_pipeline(self) -> Pipeline<Self> where Self: Send;
+
+    /// Like `into_par_pipeline()`, but also works for iterators that
+    /// are not `Send`. Requires that your code is within a Rayon
+    /// scope, which ensures that the code is already running on the
+    /// Rayon threadpool.
+    ///
+    /// Example:
+    ///
+    /// ```
+    /// use rayon::prelude::*;
+    /// use std::rc::Rc; // `Rc` is not `Send`
+    /// rayon::scope(|s| {
+    ///     let v = vec![Rc::new(1), Rc::new(2)];
+    ///     v.into_iter()
+    ///      .map(|r| *r) // iterating over i32s
+    ///      .into_par_pipeline_scoped(s)
+    ///      .find_any(|&n| n == 1);
+    /// });
+    /// ```
+    fn into_par_pipeline_scoped<'s, 'scope>(self, scope: &'s Scope<'scope>) -> PipelineScoped<'s, 'scope, Self>;
```

Comment:
> We're already a bit limited that we don't have ParallelIterator: Send, but I think perhaps we should

Yes, perhaps so.

> (On ParallelIterator, all of the methods which take closures currently require Sync but not Send, so that broadens the question considerably there.)

Probably they should be send + sync I guess.